File: src/ChromaticCircle.ts
import { Color } from "./Color";

export class ChromaticCircle {
  private readonly filter: Color;

  constructor(filter: Color) {
    this.filter = filter;
  }

  public mixColors(color: Color): Color {
    if (this.filter === color) {
      return color;
    }
    if (this.filter === Color.BLUE) {
      if (color == this.filter) {
        return color;
      }
      if (color === Color.YELLOW) {
        return Color.GREEN;
      }
      if (color === Color.RED) {
        return Color.VIOLET;
      }
    }
    if (this.filter === Color.RED) {
      if (color === Color.YELLOW) {
        return Color.ORANGE;
      }
      if (color === Color.BLUE) {
        return Color.VIOLET;
      }
    }
    if (this.filter === Color.YELLOW) {
      if (color === Color.RED) {
        return Color.ORANGE;
      }
      if (color === Color.BLUE) {
        return Color.GREEN;
      }
    }

    throw new Error(`invalid color: ${color}`);
  }
}

File: dart/lib/chromatic_circle.dart
class ChromaticCircle {
  final Color filter;

  const ChromaticCircle(this.filter);

  Color mixColors(Color color) {
    if (filter == color) {
      return color;
    }

    switch (filter) {
      case Color.blue:
        switch (color) {
          case Color.yellow:
            return Color.green;
          case Color.red:
            return Color.violet;
          default:
            throw ArgumentError('Invalid color: $color');
        }
      case Color.red:
        switch (color) {
          case Color.yellow:
            return Color.orange;
          case Color.blue:
            return Color.violet;
          default:
            throw ArgumentError('Invalid color: $color');
        }
      case Color.yellow:
        switch (color) {
          case Color.red:
            return Color.orange;
          case Color.blue:
            return Color.green;
          default:
            throw ArgumentError('Invalid color: $color');
        }
      default:
        throw ArgumentError('Invalid filter color: $filter');
    }
  }
}

enum Color {
  red,
  blue,
  yellow,
  green,
  orange,
  violet
}
